File: frontend/src/api/v3Session.ts
/**
 * DynoAI v3.0 Accelerated Calibration — API Client
 *
 * Typed functions for every /api/v3 endpoint.
 * Uses the shared axios instance from @/lib/api.
 */

import api from "@/lib/api";
import { encodePathSegment } from "@/lib/sanitize";

// ---------------------------------------------------------------------------
// Types — mirroring Python dataclasses
// ---------------------------------------------------------------------------

export type PullType = "wot_sweep" | "part_throttle" | "cruise" | "targeted";

export type PullMode = "acceleration" | "steady_state";

export interface HardwareConfig {
  engine_family: string;
  displacement_ci: number;
  cam_spec?: string;
  exhaust_type?: string;
  exhaust_brand?: string;
  air_cleaner?: string;
  throttle_body_mm?: number;
  compression_ratio?: number;
  head_work?: string;
  injector_size?: string;
  fuel_type?: string;
  altitude_ft?: number;
  tune_platform?: string;
  /** Grid override from PVV import or wizard */
  rpm_bins?: number[];
  map_bins?: number[];
}

/** Payload for creating a session; may include imported VE table to seed the GP. */
export interface CreateSessionPayload extends HardwareConfig {
  /** Optional 2D VE table from PVV import; seeds Uncertainty Map so it shows imported tune. */
  initial_ve_table?: number[][];
}

export interface PullRecommendation {
  rpm: number;
  map_kpa: number;
  gear: number;
  pull_number: number;
  pull_type: PullType;
  pull_mode?: PullMode;
  reason: string;
  expected_info_gain: number;
  remaining_uncertainty: number;
  throttle_pct: number;
  alternatives: PullRecommendation[];
}

export interface ConvergenceStatus {
  converged: boolean;
  max_uncertainty: number;
  mean_uncertainty: number;
  cells_above_threshold: number;
  total_cells: number;
  estimated_pulls_remaining: number;
}

export interface TemplateMatchInfo {
  template_id: string;
  similarity_score: number;
  is_usable: boolean;
  engine_family: string;
}

export interface SessionInitResult {
  session_id: string;
  engine_family: string;
  estimated_pulls: number;
  template_match: TemplateMatchInfo | null;
  initial_plan: PullRecommendation[];
}

export interface SessionStatus {
  session_id: string;
  state: "created" | "ready" | "in_progress" | "complete";
  engine_family: string;
  pull_count: number;
  converged: boolean;
  elapsed_s: number;
  template_match: string | null;
}

export interface PullResult {
  pull_number: number;
  observations_added: number;
  convergence: ConvergenceStatus | null;
  next_suggestion: PullRecommendation | null;
}

export interface FinalResult {
  template_id: string;
  total_pulls: number;
  session_id: string;
  session_duration_s: number;
}

export interface UncertaintyMapResult {
  ve_map: number[][];
  uncertainty_map: number[][];
  confidence_map: number[][];
  predict_time_ms: number;
  rpm_bins: number[];
  map_bins: number[];
}

export interface OverlayStatus {
  enabled: boolean;
  fuel_corrections_active: number;
  timing_corrections_active: number;
  max_fuel_correction_pct: number;
  max_timing_correction_deg: number;
  engine_family: string;
  ect_enrichment_trigger_f: number;
}

export interface TemplateListResult {
  total_templates: number;
  family: string | null;
  family_count: number;
}

export interface PullData {
  rpm: number[];
  map_kpa: number[];
  ve: number[];
}

export interface ImportVERequest {
  ve_table: number[][];
  rpm_bins: number[];
  map_bins: number[];
}

export interface ImportCorrectionsRequest {
  corrections: number[][];
  rpm_bins: number[];
  map_bins: number[];
  format: "multiplier" | "percentage";
}

export interface FinalizeRequest {
  ve_table_front: number[][];
  operator?: string;
}

export interface VetoRequest {
  rpm: number;
  map_kpa: number;
  reason?: string;
}

export interface SimulatePullRequest {
  rpm?: number;
  map_kpa?: number;
  n_points?: number;
  mode?: "quick" | "realistic";
}

export interface AFRMetrics {
  max_afr_error: number;
  mean_afr_error: number;
  data_points: number;
  zones_corrected: number;
  max_ve_correction_pct: number;
}

export interface SimulatePullResult extends PullResult {
  target_rpm: number;
  target_map_kpa: number;
  mode?: "quick" | "realistic";
  afr_metrics?: AFRMetrics | null;
}

export interface AutoSimulateRequest {
  mode?: "quick" | "realistic";
  max_pulls?: number;
}

export interface AutoSimulatePullSummary {
  pull_number: number;
  observations_added: number;
  mean_uncertainty: number | null;
}

export interface AutoSimulateResult {
  pulls_completed: number;
  converged: boolean;
  final_result: SimulatePullResult | null;
  pull_summary: AutoSimulatePullSummary[];
}

export interface MaterializeRunResult {
  success: boolean;
  run_id: string;
  session_id: string;
  materialized_at: string;
}

export interface ImportBaseVEResult {
  status: string;
  observations_added: number;
}

export interface ImportCorrectionsResult {
  status: string;
  observations_added: number;
  convergence: ConvergenceStatus | null;
  next_suggestion?: PullRecommendation | null;
}

// ---------------------------------------------------------------------------
// API functions
// ---------------------------------------------------------------------------

export async function createSession(
  payload: CreateSessionPayload
): Promise<SessionInitResult> {
  const response = await api.post<SessionInitResult>("/api/v3/session", payload);
  return response.data;
}

export async function getSession(
  sessionId: string
): Promise<SessionStatus> {
  const response = await api.get<SessionStatus>(
    `/api/v3/session/${encodePathSegment(sessionId)}`
  );
  return response.data;
}

export async function listSessions(): Promise<{ sessions: SessionStatus[] }> {
  const response = await api.get<{ sessions: SessionStatus[] }>("/api/v3/sessions");
  return response.data;
}

export async function ingestPull(
  sessionId: string,
  data: PullData
): Promise<PullResult> {
  const response = await api.post<PullResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/pull`,
    data
  );
  return response.data;
}

export async function importBaseVE(
  sessionId: string,
  data: ImportVERequest
): Promise<ImportBaseVEResult> {
  const response = await api.post<ImportBaseVEResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/import-ve`,
    data
  );
  return response.data;
}

export async function finalizeSession(
  sessionId: string,
  data: FinalizeRequest
): Promise<FinalResult> {
  const response = await api.post<FinalResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/finalize`,
    data
  );
  return response.data;
}

export async function importCorrections(
  sessionId: string,
  data: ImportCorrectionsRequest
): Promise<ImportCorrectionsResult> {
  const response = await api.post<ImportCorrectionsResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/import-corrections`,
    data
  );
  return response.data;
}

export async function suggestNextPull(
  sessionId: string
): Promise<PullRecommendation> {
  const response = await api.get<PullRecommendation>(
    `/api/v3/session/${encodePathSegment(sessionId)}/next-pull`
  );
  return response.data;
}

export async function checkConvergence(
  sessionId: string
): Promise<ConvergenceStatus> {
  const response = await api.get<ConvergenceStatus>(
    `/api/v3/session/${encodePathSegment(sessionId)}/convergence`
  );
  return response.data;
}

export async function operatorVeto(
  sessionId: string,
  data: VetoRequest
): Promise<{ status: string; rpm: number; map_kpa: number }> {
  const response = await api.post(
    `/api/v3/session/${encodePathSegment(sessionId)}/veto`,
    data
  );
  return response.data;
}

export async function getUncertaintyMap(
  sessionId: string
): Promise<UncertaintyMapResult> {
  const response = await api.get<UncertaintyMapResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/uncertainty`
  );
  return response.data;
}

export async function getOverlayStatus(
  sessionId: string
): Promise<OverlayStatus> {
  const response = await api.get<OverlayStatus>(
    `/api/v3/session/${encodePathSegment(sessionId)}/overlay`
  );
  return response.data;
}

export async function activateKillSwitch(
  sessionId: string
): Promise<{ status: string; session_id: string }> {
  const response = await api.post(
    `/api/v3/session/${encodePathSegment(sessionId)}/kill-switch`
  );
  return response.data;
}

export async function listTemplates(
  family?: string
): Promise<TemplateListResult> {
  const params = family ? { family } : {};
  const response = await api.get<TemplateListResult>("/api/v3/templates", {
    params,
  });
  return response.data;
}

export async function simulatePull(
  sessionId: string,
  data?: SimulatePullRequest
): Promise<SimulatePullResult> {
  const response = await api.post<SimulatePullResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/simulate-pull`,
    data || {}
  );
  return response.data;
}

export async function autoSimulate(
  sessionId: string,
  data?: AutoSimulateRequest
): Promise<AutoSimulateResult> {
  const response = await api.post<AutoSimulateResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/auto-simulate`,
    data || {}
  );
  return response.data;
}

export async function materializeRun(
  sessionId: string
): Promise<MaterializeRunResult> {
  const response = await api.post<MaterializeRunResult>(
    `/api/v3/session/${encodePathSegment(sessionId)}/materialize-run`
  );
  return response.data;
}

File: api/routes/v3_session.py
"""
DynoAI v3.0 Session API Routes
=================================

REST endpoints for the Accelerated Calibration Platform.

Prefix: ``/api/v3``
"""

from __future__ import annotations

import logging

from flask import Blueprint, jsonify, request

from api.errors import NotFoundError, ValidationError, with_error_handling

logger = logging.getLogger(__name__)

v3_session_bp = Blueprint("v3_session", __name__, url_prefix="/api/v3")


# ---------------------------------------------------------------------------
# Session lifecycle
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session", methods=["POST"])
@with_error_handling
def create_session():
    """Create and initialize a new v3 tuning session."""
    data = request.get_json()
    if not data:
        raise ValidationError("Request body required (hardware config)")

    if "engine_family" not in data:
        raise ValidationError("engine_family is required")

    from api.services.v3_session_service import create_session as _create
    result = _create(data)
    return jsonify(result), 201


@v3_session_bp.route("/session/<session_id>", methods=["GET"])
@with_error_handling
def get_session(session_id: str):
    """Get session status."""
    from api.services.v3_session_service import get_session as _get
    try:
        result = _get(session_id)
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


@v3_session_bp.route("/sessions", methods=["GET"])
@with_error_handling
def list_sessions():
    """List all active v3 sessions."""
    from api.services.v3_session_service import list_sessions as _list
    return jsonify({"sessions": _list()})


# ---------------------------------------------------------------------------
# Pull ingestion
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session/<session_id>/pull", methods=["POST"])
@with_error_handling
def ingest_pull(session_id: str):
    """Ingest pull data into a session."""
    data = request.get_json()
    if not data:
        raise ValidationError("Request body required (pull data)")

    for field in ("rpm", "map_kpa", "ve"):
        if field not in data:
            raise ValidationError(f"'{field}' array is required")

    from api.services.v3_session_service import ingest_pull as _ingest
    try:
        result = _ingest(
            session_id,
            rpm=data["rpm"],
            map_kpa=data["map_kpa"],
            ve=data["ve"],
        )
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


# ---------------------------------------------------------------------------
# Imports (base VE + corrections)
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session/<session_id>/import-ve", methods=["POST"])
@with_error_handling
def import_base_ve(session_id: str):
    """Import a base VE table and seed the GP surrogate."""
    data = request.get_json()
    if not data:
        raise ValidationError("Request body required (base VE import)")

    for field in ("ve_table", "rpm_bins", "map_bins"):
        if field not in data:
            raise ValidationError(f"'{field}' is required")

    from api.services.v3_session_service import import_base_ve as _import
    try:
        result = _import(
            session_id,
            ve_table=data["ve_table"],
            rpm_bins=data["rpm_bins"],
            map_bins=data["map_bins"],
        )
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


@v3_session_bp.route("/session/<session_id>/import-corrections", methods=["POST"])
@with_error_handling
def import_corrections(session_id: str):
    """Import correction grid into an active session."""
    data = request.get_json()
    if not data:
        raise ValidationError("Request body required (corrections import)")

    for field in ("corrections", "rpm_bins", "map_bins", "format"):
        if field not in data:
            raise ValidationError(f"'{field}' is required")

    from api.services.v3_session_service import import_corrections as _import
    try:
        result = _import(
            session_id,
            corrections=data["corrections"],
            rpm_bins=data["rpm_bins"],
            map_bins=data["map_bins"],
            fmt=data["format"],
        )
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


# ---------------------------------------------------------------------------
# Finalize
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session/<session_id>/finalize", methods=["POST"])
@with_error_handling
def finalize_session(session_id: str):
    """Finalize session and store template."""
    data = request.get_json()
    if not data or "ve_table_front" not in data:
        raise ValidationError("ve_table_front is required")

    from api.services.v3_session_service import finalize_session as _finalize
    try:
        result = _finalize(
            session_id,
            ve_table_front=data["ve_table_front"],
            operator=data.get("operator", "unknown"),
        )
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


@v3_session_bp.route("/session/<session_id>/materialize-run", methods=["POST"])
@with_error_handling
def materialize_run(session_id: str):
    """Materialize latest v3 correction grid as a run artifact for /api/apply."""
    from api.services.v3_session_service import materialize_run as _materialize
    try:
        result = _materialize(session_id)
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


# ---------------------------------------------------------------------------
# Pull advisor
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session/<session_id>/next-pull", methods=["GET"])
@with_error_handling
def suggest_next_pull(session_id: str):
    """Get the advisor's next pull recommendation."""
    from api.services.v3_session_service import suggest_next_pull as _suggest
    try:
        result = _suggest(session_id)
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


@v3_session_bp.route("/session/<session_id>/convergence", methods=["GET"])
@with_error_handling
def check_convergence(session_id: str):
    """Get convergence status."""
    from api.services.v3_session_service import check_convergence as _check
    try:
        result = _check(session_id)
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


@v3_session_bp.route("/session/<session_id>/veto", methods=["POST"])
@with_error_handling
def operator_veto(session_id: str):
    """Operator veto a suggested operating point."""
    data = request.get_json()
    if not data or "rpm" not in data or "map_kpa" not in data:
        raise ValidationError("rpm and map_kpa are required")

    from api.services.v3_session_service import operator_veto as _veto
    try:
        result = _veto(
            session_id,
            rpm=data["rpm"],
            map_kpa=data["map_kpa"],
            reason=data.get("reason", ""),
        )
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


# ---------------------------------------------------------------------------
# GP surrogate
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session/<session_id>/uncertainty", methods=["GET"])
@with_error_handling
def get_uncertainty_map(session_id: str):
    """Get GP uncertainty map for the session."""
    from api.services.v3_session_service import get_uncertainty_map as _unc
    try:
        result = _unc(session_id)
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


# ---------------------------------------------------------------------------
# Overlay
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session/<session_id>/overlay", methods=["GET"])
@with_error_handling
def get_overlay_status(session_id: str):
    """Get bounded overlay status."""
    from api.services.v3_session_service import get_overlay_status as _overlay
    try:
        result = _overlay(session_id)
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


@v3_session_bp.route("/session/<session_id>/kill-switch", methods=["POST"])
@with_error_handling
def kill_switch(session_id: str):
    """Activate overlay kill switch."""
    from api.services.v3_session_service import kill_switch as _kill
    try:
        result = _kill(session_id)
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


# ---------------------------------------------------------------------------
# Templates
# ---------------------------------------------------------------------------

@v3_session_bp.route("/templates", methods=["GET"])
@with_error_handling
def list_templates():
    """List templates in the library."""
    family = request.args.get("family")
    from api.services.v3_session_service import list_templates as _list
    return jsonify(_list(engine_family=family))


# ---------------------------------------------------------------------------
# Simulation
# ---------------------------------------------------------------------------

@v3_session_bp.route("/session/<session_id>/simulate-pull", methods=["POST"])
@with_error_handling
def simulate_pull(session_id: str):
    """
    Generate and ingest simulated pull data.

    Body (all optional):
        mode:     "quick" (random scatter) or "realistic" (DynoSimulator + AutoTune)
        rpm:      Target RPM (defaults to advisor suggestion)
        map_kpa:  Target MAP (defaults to advisor suggestion)
        n_points: Points for quick mode (default 8)
    """
    data = request.get_json() or {}
    mode = data.get("mode", "quick")

    try:
        if mode == "realistic":
            from api.services.v3_session_service import simulate_pull_realistic as _sim
            result = _sim(
                session_id,
                rpm=data.get("rpm"),
                map_kpa=data.get("map_kpa"),
            )
        else:
            from api.services.v3_session_service import simulate_pull as _sim
            result = _sim(
                session_id,
                rpm=data.get("rpm"),
                map_kpa=data.get("map_kpa"),
                n_points=data.get("n_points", 8),
            )
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)
    return jsonify(result)


@v3_session_bp.route("/session/<session_id>/auto-simulate", methods=["POST"])
@with_error_handling
def auto_simulate(session_id: str):
    """
    Auto-run simulate pulls until converged or max_pulls reached.

    Body (all optional):
        mode:      "quick" or "realistic" (default "realistic")
        max_pulls: Maximum number of pulls (default 25)
    """
    data = request.get_json() or {}
    mode = data.get("mode", "realistic")
    max_pulls = min(data.get("max_pulls", 25), 50)  # cap at 50

    from api.services.v3_session_service import (
        simulate_pull as _sim_quick,
        simulate_pull_realistic as _sim_real,
        check_convergence as _check,
    )

    results = []
    try:
        for i in range(max_pulls):
            if mode == "realistic":
                result = _sim_real(session_id)
            else:
                result = _sim_quick(session_id)

            results.append(result)

            # Check convergence
            conv = result.get("convergence")
            if conv and conv.get("converged"):
                break
    except KeyError:
        raise NotFoundError(resource="V3 Session", identifier=session_id)

    return jsonify({
        "pulls_completed": len(results),
        "converged": results[-1].get("convergence", {}).get("converged", False) if results else False,
        "final_result": results[-1] if results else None,
        "pull_summary": [
            {
                "pull_number": r.get("pull_number"),
                "observations_added": r.get("observations_added"),
                "mean_uncertainty": r.get("convergence", {}).get("mean_uncertainty") if r.get("convergence") else None,
            }
            for r in results
        ],
    })
